src/rayleigh_quotient.jl

```diff
@@ -80,73 +116,168 @@ function solve(prob::ConstrainedRayleighQuotientProblem, alg::RQ_CHOL)
+    bk = first(b[inds_remove]) # only one element, no support for b Span
+    Ck = C[inds_remove, :] # replace mult by J with slicing, see above
+    return sparse((C - (1/bk) * b * Ck)[inds_keep, :])
+end
+
+function _inds_to_remove_and_keep(b, N)
+    inds_remove = partialsortperm(eachrow(b), 1:size(b, 2); by=norm, rev=true)
+    inds_keep = map(!in(inds_remove), 1:N)
+    return inds_remove, inds_keep
+end
+
+function _solve_homo_prob(C_reduced, prob::ConstrainedRayleighQuotientProblem, alg)
+    b = _get_b(prob)
+    C = prob.C
+    homo_prob = ConstrainedRayleighQuotientProblem(prob.Q, C_reduced, zero(b))
+    sol = solve(homo_prob, RQ_HOMO(alg.krylov_howmany, alg.krylov_kwargs))
+    t = dot(b, C, sol) / dot(b, b)
+    return sol / t
+end
+
+function _solve_homo_prob(augC, prob::ConstrainedRayleighQuotientProblem{Q,Cmat,<:Span} , alg) where {Q,Cmat}
+    b = _get_b(prob)[:, 1] # for b Span, normalization is not specified
     homo_prob = ConstrainedRayleighQuotientProblem(prob.Q, augC, zero(b))
-    sol = solve(homo_prob, RQ_HOMO(alg.eps))
-    z = dot(b, C, sol) / dot(b, b)
-    return sol / z
+    return solve(homo_prob, RQ_HOMO(alg.krylov_howmany, alg.krylov_kwargs))
 end
 
-function solve(prob::ConstrainedRayleighQuotientProblem, alg::RQ_HOMO)
+function solve(prob::ConstrainedRayleighQuotientProblem{Q,Cmat,<:AbstractVector}, alg::RQ_HOMO) where {Q,Cmat}
     @assert iszero(prob.b) "b must be zero"
     C = prob.C
-    P = I - C' * (factorize(Hermitian(C * C')) \ C)
-    eig = eigen(Hermitian(P' * prob.Q.quadratic_form * P))
-    _select_vectors(eig.vectors, P, prob.b, alg.eps)
-end
-
-function _select_vectors(eigvecs, P, b::AbstractMatrix, eps=DEFAULT_SELECTVEC_EPS)
-    vecs = eachcol(eigvecs)
-    v = similar(first(vecs))
-    f = x -> any(>(eps) ∘ abs, mul!(v, P, x))
-    i1 = findfirst(f, vecs)
-    vecinds = [i1]
-    for n in 2:size(b, 2)
-        push!(vecinds, findnext(f, vecs, last(vecinds) + 1))
-    end
-    return reduce(hcat, vecs[vecinds])
+    P, PQP = _get_P_PQP_homo(C, prob.Q.quadratic_form)
+    howmany = 1
+    _, eigvecs, info = eigsolve(PQP, size(C, 2), howmany, :SR; ishermitian=true)
+    ind = _select_vector_index(eigvecs, P)
+    !isnothing(ind) && return eigvecs[ind] # if an ok vector is found, return it, otherwise, look for more
+    #=howmany = size(C, 1) + 1 # this is often too large=#
+    _, eigvecs, info = eigsolve(PQP, first(eigvecs), alg.krylov_howmany, :SR; ishermitian=true, alg.krylov_kwargs...)
+    ind = _select_vector_index(eigvecs, P)
+    !isnothing(ind) && return eigvecs[ind]
+    error("Solution was not found. Increase alg.krylov_howmany and/or krylovdim in alg.krylov_kwargs to look for more candidates")
 end
 
+function _get_P_PQP_homo(C, Q)
+    #=P = I - C' * InverseMap(factorize(Hermitian(C * C'))) * C=#
+    P = I - C' * InverseMap(lu(Hermitian(C * C'))) * C # sparse cholesky is incompatible with ldiv!
+    return P, P' * Q * P
```

Comment:
We should let LinearMaps know that this is Hermitian, so that krylovkit can select a lanczos algorithm. Something like `LinearMap(P, ishermitian=true)` might work.

Reply:
We supply ishermitian=true to eigsolve so it should know about it. But it might make sense to do it at the level of LinearMaps. Maybe `LinearMap(P' * Q * P, ishermitian=true)` then, since this is the matrix we provide to eigsolve?

Reply:
Both should be hermitian

Reply:
Turns out that KrylovKit doesn't check if the matrix is hermitian if it is not an AbstractMatrix, so I think we have to supply it to eigsolve.